File: code/ch07/java/ObserverClinic.java
// §7.2.2 Observers and Subscribers — a subject that notifies without knowing
// its observers' concrete types (clinic scheduler: waiting-room display).
// Compact source file (JDK 25+): run with `java ObserverClinic.java`.
interface Observer { void update(Appointment appointment); }

class Appointment {
  String status = "booked";
  private final List<Observer> observers = new ArrayList<>();

  void subscribe(Observer observer) { observers.add(observer); }

  void setStatus(String newStatus) {
    status = newStatus;
    for (Observer obs : observers) obs.update(this);   // any Observer will do
  }
}

class WaitingRoomDisplay implements Observer {
  public void update(Appointment appointment) {
    IO.println("display: appointment is now " + appointment.status);
  }
}

void main() {
  Appointment appt = new Appointment();
  appt.subscribe(new WaitingRoomDisplay());
  appt.setStatus("arrived");             // prints: display: appointment is now arrived
}

File: code/ch07/python/observer_clinic.py
# §7.2.2 Observers and Subscribers — a subject that notifies without knowing
# its observers' concrete types (clinic scheduler: waiting-room display).
class Appointment:
  def __init__(self, status="booked"):
    self.status = status
    self._observers = []

  def subscribe(self, observer):
    self._observers.append(observer)

  def set_status(self, new_status):
    self.status = new_status
    for obs in self._observers:      # any object with update() will do
      obs.update(self)

class WaitingRoomDisplay:
  def update(self, appointment):
    print(f"display: appointment is now {appointment.status}")

appt = Appointment()
appt.subscribe(WaitingRoomDisplay())
appt.set_status("arrived")               # prints: display: appointment is now arrived
